### PR TITLE
Task7 post a comment to an article

### PR DESCRIPTION
**Pull request by LBerry123**
Added the endpoint to post a comment to an article

### CODE REVIEW

**Comment by Spoomgull** [comment]
i think i would like to wait until task 14 to refactor unless you find something that needs to be fixed

File: __tests__/apiTests.test.js [hunk 1]
```diff
@@ -164,4 +164,45 @@ describe("GET /api/articles/:article_id/comments",()=>{
                 expect(body.msg).toBe("Invalid query params!!")
             })
     })
+})
+
+describe("POST /api/articles/:article_id/comments",()=>{
+    test("get status code 200 and responds with the posted comment",()=>{
+        const comment = {username: "jonny",body: "This article is pretty great! :)"}
```
Comment (roisinoreilly) [inline __tests__/apiTests.test.js:171]: did you update the database to include this username?

File: __tests__/apiTests.test.js [hunk 2]
```diff
@@ -164,4 +164,45 @@ describe("GET /api/articles/:article_id/comments",()=>{
...
+        return request(app)
+
+            .post("/api/articles/2/comments")
+            .send(comment)
+            .expect(201)
+                .then(({body})=>{
+                    console.log(body)
```
Comment (roisinoreilly) [inline __tests__/apiTests.test.js:178]: don't forget to remove any console.logs before pushing up finished code

File: __tests__/apiTests.test.js [hunk 3]
```diff
@@ -164,4 +164,45 @@ describe("GET /api/articles/:article_id/comments",()=>{
...
+                    const postedComment = body.comment
+                    expect(postedComment).toMatchObject([{"article_id": 2, "author": "butter_bridge", "body": "This article is pretty great! :)", "comment_id": 19, "votes": 0}])
```
Comment (roisinoreilly) [inline __tests__/apiTests.test.js:180]: this username doesn't match the one in your post request itself?

File: __tests__/apiTests.test.js [hunk 4]
```diff
@@ -164,4 +164,45 @@ describe("GET /api/articles/:article_id/comments",()=>{
...
+                    expect(typeof postedComment[0].created_at).toBe("string")
```
Comment (roisinoreilly) [inline __tests__/apiTests.test.js:181]: you could include created_at in the toMatchObject and instead do something like created_at: expect.any(String) or expect.toBeDateString()

File: app/model.js [hunk 5]
```diff
@@ -22,4 +22,28 @@ exports.selectAllArticles = ()=>{
...
+}
+
+exports.updateComments = (comment,id)=>{
+
+    const {body} = comment
+    const {article_id} = id
+    const author = comment.username
+ 
+    if(body===undefined||article_id===undefined||author===undefined){
+        console.log("hello")
```
Comment (roisinoreilly) [inline app/model.js:34]: remove this

File: app/model.js [hunk 6]
```diff
@@ -22,4 +22,28 @@ exports.selectAllArticles = ()=>{
...
\ No newline at end of file
+}
+
+exports.updateComments = (comment,id)=>{
+
+    const {body} = comment
+    const {article_id} = id
+    const author = comment.username
+ 
+    if(body===undefined||article_id===undefined||author===undefined){
```
Comment (roisinoreilly) [inline app/model.js:33]: a more readable way to write this would be `if (!body || !article_id || !author)`

File: app/model.js [hunk 7]
```diff
@@ -22,4 +22,28 @@ exports.selectAllArticles = ()=>{
...
+        console.log("hello")
+       return err
+    }
+
+    return db.query(`SELECT username FROM users WHERE name =$1`,[author]).then(({rows})=>{
```
Comment (roisinoreilly) [inline app/model.js:38]: you shouldn't be querying the database for a name here, the posted comment should just be a valid username in the database, no need to use names

File: app/model.js [hunk 8]
```diff
@@ -22,4 +22,28 @@ exports.selectAllArticles = ()=>{
...
+        if(rows.length===0){return Promise.reject({status:400,msg:"help"})}
```
Comment (roisinoreilly) [inline app/model.js:39]: this isn't a useful error message for the client

File: app/controller.js [hunk 9]
```diff
@@ -59,8 +58,22 @@ exports.getComments = (req,res,next) =>{
       
         return res.status(200).send({comments:rows})
 
-    }).catch(({code})=>{if(code==="22P02"){
-        return res.status(400).send({msg:"Invalid query params!!"})
-        }else{code}
+    }).catch(({code})=>{
```
Comment (roisinoreilly) [inline app/controller.js:61]: this should be error

File: app/controller.js [hunk 10]
```diff
@@ -59,8 +58,22 @@ exports.getComments = (req,res,next) =>{
...
+}
+
+exports.postComment = (req,res,next) =>{
+    const comment = req.body
+    const articleId = req.params
+    
+    updateComments(comment,articleId).then(({rows})=>{
+        return res.status(201).send({comment:rows})
+    }).catch(({code})=>{
+        if(code === undefined){return res.status(400).send({msg:"Invalid username"})}
```
Comment (roisinoreilly) [inline app/controller.js:74]: this should also be error not code, and you don't need to check if it's undefined here, just pass the error to your error handling middleware to deal with please!

File: app/model.js (hunk 5 above)
Comment (roisinoreilly) [inline app/model.js:34]: remove this

**Review by roisinoreilly — COMMENTED** [review]
I've left quite a few comments on this PR for you to look through. The main concerns I have with this ticket is that your controller doesn't need to decipher through error codes, just pass the errors to your handler, and also that you're allowing a post request with a name instead of a username, so there's functionality in the model to turn a name into a username that doesn't need to be there. Remove that extra query and just test with a valid username, and have a look at tidying up the code a bit with the console.logs that have been left in. Any questions let me know

**Pull request merged**